service/history/historyCache.go

```diff
@@ -124,12 +130,15 @@ func (c *historyCache) getOrCreateWorkflowExecution(domainID string,
 				// TODO see issue #668, there are certain type or errors which can bypass the clear
 				context.clear()
 			}
-			context.Unlock()
+			context.locker.Unlock()
 			c.Release(key)
 		}
 	}
 
-	context.Lock()
+	if err := context.locker.Lock(ctx); err != nil {
+		// ctx is done before lock can be acquired
```

Comment:
missing "c.Release(key)" ?
when the context is acquired after line 124, there is a internal counter in the caching layer which will be ++ed.